Add workflow invocation detail component

mvdbeek opened the pull request:
- Shows workflow inputs and outputs (parameters + data)
- Shows all workflow steps (even the steps that haven't been scheduled yet)
- Shows subworkflow invocations inline
- Shows a table of all jobs for an invocation step
- Shows parameters for a job
- Shows dependencies of a job (why hasn't this step started yet ?)
 
<img width="1424" alt="Screenshot 2020-12-20 at 19 00 40" src="https://user-images.githubusercontent.com/6804901/102720608-ace45d00-42f5-11eb-9adc-e5685d811482.png">

TODO:
  - [x] Is the provider pattern the right thing to use ?
  - [x] Should data pass through pouchdb ?
  - [x] History panel should update ?
  - [x] How do I modify/delete objects ?
  - [ ] Indentation can be a bit confusing
  - [ ] Maybe collect datasets / dataset collections as data inputs & data outputs ?

mvdbeek committed:
Get collection view in getRecentInvocations

Instead of the the default element view. We don't use the returned extra
data.

mvdbeek committed:
Include invocation step parameters in invocation elements API

And use the smaller collection view.
Also fixes finding `data_input` and `data_collection_input` steps, which
relied on step_type tool for finding invocation inputs.

mvdbeek committed:
Add a workflow invocation details component

Now ... which components can I actually reuse ? :laughing:

mvdbeek committed:
WIP render output via masons component

mvdbeek committed:
More experiments

mvdbeek committed:
Make dataset collections display in workflow invocations menu

mvdbeek committed:
Replace Dataset/index and DatasetCollection/index with HistoryContentItem

Nerdinacan committed:
Added sample Datasetprovider for marius

mvdbeek committed:
Call load on created() in DatasetProvider

mvdbeek committed:
Use DatasetProvider in WorkflowOutputs

mvdbeek committed:
Display invocation parameters a table

mvdbeek committed:
Track DatasetUIWrapper.vue

mvdbeek committed:
Create DatasetCollectionProvider

mvdbeek committed:
Move handling of expanded into content wrapper components

mvdbeek committed:
Fix watching

mvdbeek committed:
Add DatasetCollectionContentProvider

mvdbeek committed:
Add invocation output collectio display

mvdbeek committed:
Indent nested components

mvdbeek committed:
Try overflow scroll on container

dannon committed:
Deactivate responsivity on invocations table (prevents rows from expanding out of control -- overflow works)

mvdbeek committed:
Pass item_data objects, dispatch components on type

This is much less verbose

mvdbeek committed:
Drop log

mvdbeek committed:
Label data inputs and outputs

mvdbeek committed:
Add a toolStore, use it to get tool names

mvdbeek committed:
Break out WorkflowStep component

mvdbeek committed:
Fetch & show invocation step

mvdbeek committed:
Stop verbose console errors when vnode.datta is undefined

mvdbeek committed:
Display job parameters in Job table

mvdbeek committed:
Use JobDetails to render Job parameters

mvdbeek committed:
Give me all the details

mvdbeek committed:
Add Job Information, Subworkflow steps

Resolved conflicts from 8e6cc9b by picking upstream component,
but should go back and add missing functionality (stdout/stderr inline,
pass in objects instead of ids).

mvdbeek committed:
Fix subworkflow step display

mvdbeek committed:
Fetch workflow instance when necessary

mvdbeek committed:
Display invocation steps using workflow instead of invocation

This means we can display steps before an invocation is actually
scheduled.

mvdbeek committed:
Reload recent invocations when cancelling invocation

mvdbeek committed:
Make new-ish invocations cancelable

mvdbeek committed:
Rebase me into miniminal solution

mvdbeek committed:
Poll for WorkflowInvocationStep

mvdbeek committed:
Tweak input module data display

mvdbeek committed:
Keep polling invocation state until jobs are terminal

If there are any.

mvdbeek committed:
model fixes rebase

mvdbeek committed:
better pollUntil logic

mvdbeek committed:
Rename StepJobs to JobStep, unifies step components

mvdbeek committed:
Don't need lodash for this

mvdbeek committed:
use bind shorthand

mvdbeek committed:
MOre step state fixes and linting

mvdbeek committed:
Prettier

mvdbeek committed:
Vue linting fixes

mvdbeek committed:
Fix invocation api endpoint option parsing

mvdbeek committed:
Run prettier again

mvdbeek committed:
Listen to changes in datasetsStore

This seems like a reasonably efficient way of keeping datasets synced.

jmchilton commented on client/src/components/JobInformation/JobInformation.vue:
```diff
@@ -88,79 +44,34 @@
-        hda_id: {
-            type: String,
-            required: true,
-        },
```
Past John convinced himself there were times the Job API would reject requests for information but that if the an accessible HDA was passed it made sense to render the information. There must be some comments about it server side somewhere? I hope you have thought through all of that. Like maybe this component is sitting on a published Page (it can be used from the Markdown engine) with a public HDA that doesn't belong to the user but that the viewer of the page should be able to see? Maybe this wouldn't work in that case anyway, no promises either way.

mvdbeek replied:
The github UI didn't get updated, but this was merged with https://github.com/galaxyproject/galaxy/pull/10980. I'll see what happens. In case this doesn't work, do you think we can consider a job accessible if any one of its outputs is accessible ?

jmchilton replied:
Who merged that without my review.... ... oh 🤣 .

Umm.. I think effectively it was - if you want to formalize that by just allowing access by job_id I think that is fine. I was clearly being very conservative here - because it is a thorny question and I didn't want to make any decisions, just preserve what was working.

mvdbeek committed:
Disable CacheWorker during unit tests

mvdbeek committed:
Also sync updates for dataset collections

Nerdinacan committed:
removed unnecessary test build check in worker definition

Nerdinacan committed:
fixed import locations for caching module

Nerdinacan committed:
removed blank file, exported component defs without invoking global Vue

Nerdinacan committed:
fixed cache module imports in new store modules

Nerdinacan committed:
move components to workflow invocation folder

Nerdinacan committed:
implemented sample dataset provider

Nerdinacan committed:
replaced v-bind on single vars

Nerdinacan committed:
squash

mvdbeek committed:
Cleanup imports, remove unused providers

mvdbeek committed:
Use subscriptions in providers

This either fetches items from the server or uses the contents of pouch
and watches for changes in pouch.
Also adapts DatasetInformation.vue to use the new DatasetProvider.

mvdbeek committed:
Avoid global eventHub for selectCollection

This prevents  opening a collection in the history panel when expanding
a collection in the invocation view, which I think is confusing.

mvdbeek committed:
Don't allow editing item name for collection elements

And don't add hidden/deleted to title, that is already reflected in the
icons.

mvdbeek committed:
Use invocationStepProvider to avoid state weirdnesses

mvdbeek committed:
Don't poll admin endpoint as regular user

mvdbeek committed:
Simplify check for invocation data

mvdbeek committed:
Add subworkflow icon (fa-tasks), make jobs expandable like all other step properties

mvdbeek committed:
Poll history while invocation is active

This keeps the dataset and dataset collection states up to date (for as
far as update_time is set correctly, this doesn't currently work for the
beta history panel either).

mvdbeek committed:
Dataset / element tests

mvdbeek committed:
Prettier

mvdbeek committed:
Tests for RecentInvocations event handling

mvdbeek committed:
Restore limit

And then replace it with pagination asap

mvdbeek committed:
Add Invocations unit tests

mvdbeek committed:
Add DatasetUIWrapper test

mvdbeek committed:
Add DatasetCollectionUIWrapper unit tests

mvdbeek committed:
Improve invocation state display for jobless steps

mvdbeek committed:
Remove unused props

mvdbeek committed:
Simplify WorkflowInvocationState and add test

mvdbeek committed:
Use Workflow to find version of StoredWorkflow

This returns the "correct" StoredWorkflow for a workflow instance id.
The WorklowInvocationDetails component uses the workflow to display
steps that haven't been scheduled yet. Without this we may be displaying
the wrong steps.

mvdbeek committed:
Don't repeat the word invocation so much ...

mvdbeek committed:
Only stop polling when scheduling and jobs are finished

mvdbeek committed:
Rebase fixes

mvdbeek committed:
Fix getter access

mvdbeek committed:
Poll once after invocation jobs are terminal

mvdbeek committed:
Only poll for updates, not exisiting datasets in history

mvdbeek committed:
Fix wrong variable use

mvdbeek committed:
Fix WorkflowInvocationState test for new job state logic

mvdbeek committed:
Not using filter anymore

mvdbeek committed:
Add test for monitorHistoryUntilTrue

mvdbeek committed:
Add test for invocationStepMonitor
